File: scripts/bin/vyatta-show-vrrp.pl
#!/usr/bin/perl
# 
# **** License ****
# 
# Copyright (c) 2019 AT&T Intellectual Property.
# All rights reserved.
# 
# Copyright (c) 2014-2015 by Brocade Communications Systems, Inc.
# All rights reserved.
#
# This code was originally developed by Vyatta, Inc.
# Portions created by Vyatta are Copyright (C) 2007-2012 Vyatta, Inc.
# All Rights Reserved.
# 
# Author: John Southworth
# Date: May 2012  
# Description: Process operational data from keepalived
# 
# SPDX-License-Identifier: GPL-2.0-only
#
# **** End License ****
#

use strict;
use warnings;
use lib "/opt/vyatta/share/perl5";

use Getopt::Long;
use Vyatta::VRRP::OPMode;
use Vyatta::Keepalived qw(is_running);
use Sort::Versions;
use v5.10;

my ($show, $intf, $vrid, $sync);
my $quiet = "";
GetOptions("show=s" => \$show,
           "intf=s" => \$intf,
           "vrid=s" => \$vrid,
           "sync=s" => \$sync,
           "quiet"  => \$quiet
          );

sub list_vrrp_intf {
  my $intf = shift; 
  my $hash = {};
  process_data $hash;
  if ($intf) {
    printf "%s\n", join " ", sort {versioncmp($a, $b)} keys(%{$hash->{instances}->{$intf}});
  } else {
    printf "%s\n", join " ", sort {versioncmp($a, $b)} keys(%{$hash->{instances}});
  }
}

sub list_vrrp_groups {
  my $intf = shift; 
  my $group = {};
  my $hash = {};
  process_data $hash;
  foreach my $group (sort {versioncmp($a, $b)} keys(%{$hash->{instances}->{$intf}})) {
    printf "%s\n", join " ", sort {versioncmp($a, $b)} $group;
  }
}

sub list_vrrp_sync_groups {
    my $hash = {};
    process_data $hash;
    printf "%s\n", join " ", sort {versioncmp($a, $b)} keys(%{$hash->{'sync-groups'}});
}

sub show_vrrp_summary {
  my ($intf, $vrid) = @_;
  my $hash = {};
  process_data $hash;
  return if (check_intf($hash, $intf, $vrid));
  print_summary $hash, $intf, $vrid;
}

sub show_vrrp_interfaces {
  my ($intf, $vrid) = @_;
  my $hash = {};
  process_data $hash;
  return if (check_intf($hash, $intf, $vrid));
  print_interfaces $hash, $intf, $vrid;
}

sub show_vrrp_stats {
  my ($intf, $vrid) = @_;
  my $hash = {};
  process_stats $hash;
  return if (check_intf($hash, $intf, $vrid));
  print_stats $hash, $intf, $vrid;
}

sub show_vrrp_detail {
  my ($intf, $vrid) = @_;
  my $hash = {};
  process_data $hash;
  return if (check_intf($hash, $intf, $vrid));
  print_detail $hash, $intf, $vrid;
}

sub show_vrrp_sync_groups {
  my $sync = shift;
  my $hash = {};
  process_data $hash;
  if ($sync && !exists($hash->{'sync-groups'}->{$sync})){
    print "Sync-group: $sync does not exist\n";
    return;
  }
  print_sync $hash, $sync;
}

if (!Vyatta::Keepalived::is_running()) {
    Vyatta::VRRP::OPMode::vrrp_norun() if !$quiet;
    exit;
}
if ($show eq 'summary') {
    show_vrrp_summary $intf, $vrid;
} elsif ($show eq 'interfaces') {
    show_vrrp_interfaces $intf, $vrid;
} elsif ($show eq 'detail') {
    show_vrrp_detail $intf, $vrid;
} elsif ($show eq 'stats') {
    show_vrrp_stats $intf, $vrid;
} elsif ($show eq 'sync') {
    show_vrrp_sync_groups $sync;
} elsif ($show eq 'interface') {
    list_vrrp_intf $intf;
} elsif ($show eq 'syncs') {
    list_vrrp_sync_groups;
} elsif ($show eq 'groups') {
    list_vrrp_groups $intf;
} else{
    exit;
}

File: scripts/bin/vyatta-show-vrrp.py
#! /usr/bin/python3

# Copyright (c) 2020 AT&T Intellectual Property.
# All rights reserved.
#
# SPDX-License-Identifier: GPL-2.0-only

import argparse
from typing import Dict

import vci
import vyatta.show_vrrp_cmds as vrrp_show
from vyatta.vyatta_vrrp_vci import State
import vyatta.keepalived.config_file as impl_conf
import vyatta.keepalived.util as util
import vyatta.keepalived.dbus.process_control as process_control
import vyatta.abstract_vrrp_classes as abstract_impl


def process_arguments(command: str, intf: str, vrid: str, sync: str) -> str:
    show_output: str = "VRRP is not running"
    if command == "summary":
        keepalived_implementation: abstract_impl.ConfigFile \
            = impl_conf.KeepalivedConfig()
        current_state = State(keepalived_implementation)
        state_dict = current_state.get()
        if state_dict == {}:
            return show_output
        show_output = vrrp_show.show_vrrp_summary(current_state.get())
    else:
        pc = process_control.ProcessControl()
        if not pc.is_running():
            return show_output
        file_contents: str
        json_repr: Dict
        if command == "detail" or command == "interface" or command == "sync":
            keepalived_responding = pc.dump_keepalived_data()
            if not keepalived_responding:
                return "Keepalived is not responding"
            with open(util.KEEPALIVED_DATA_FILE_PATH, "r") as file_obj:
                file_contents = file_obj.read()
            json_repr = vrrp_show.convert_data_file_to_dict(file_contents)
            if command == "detail":
                show_output = vrrp_show.show_vrrp_detail(json_repr)
            elif command == "interface":
                show_output = vrrp_show.show_vrrp_interface(json_repr, intf, vrid)
            elif command == "sync":
                show_output = vrrp_show.show_vrrp_sync(json_repr, sync)
            else:
                print("Something has gone horribly wrong")
        else:
            keepalived_responding = pc.dump_keepalived_stats()
            if not keepalived_responding:
                return "Keepalived is not responding"
            with open(util.KEEPALIVED_STATS_FILE_PATH, "r") as file_obj:
                file_contents = file_obj.read()
            json_repr= vrrp_show.convert_stats_file_to_dict(file_contents)
            show_output = vrrp_show.show_vrrp_statistics_filters(
                json_repr, intf, vrid
            )
    return show_output


def main() -> str:
    parser = argparse.ArgumentParser(description="Show output for VRRP")
    parser.add_argument("show" , 
        help="""The show command to display\n
        summary - show vrrp\n
        detail - show vrrp detail\n
        interface - show vrrp interface\n
        stats - show vrrp statistics\n
        sync - show vrrp sync-group\n
        """,
        choices=["summary", "detail", "interface", "stats", "sync"]
    )
    parser.add_argument(
        "--intf", help="Filter on interface", default=""
    )
    parser.add_argument(
        "--vrid", help="Filter on group", default=""
    )
    parser.add_argument(
        "--sync", help="Filter sync group", default=""
    )
    args = parser.parse_args()
    command: str = args.show
    filter_intf: str = args.intf
    filter_vrid: str = args.vrid
    filter_sync: str = args.sync
    return process_arguments(command, filter_intf, filter_vrid, filter_sync)


if __name__ == "__main__":
    main()
